### ISSUE TITLE
TSX file indentation

### ISSUE THREAD

**Issue by tbekolay** [ananthakumaran/tide#78]
First, thanks a lot for tide! It's been super helpful as I've been dipping my toes into the world of modern web development.

When writing files with [JSX](https://www.typescriptlang.org/docs/handbook/jsx.html) syntax, it would be ideal to increase the indent level at opening tags (`<div>`) and decrease the indent level at closing tags (`</div>`). Currently, all tags will end up at the same level, which hurt readability.

How difficult would this be to implement? If it's a relatively big job, I'm happy to attempt to implement this myself, but would appreciate any pointers to the places in the current code that you think it might touch. I suspect that it might be best to implement as a new major mode that delegates most things to `typescript-mode`, as then it can be associated with the `.tsx` file extension, but again would appreciate any guidance you can provide!

**Comment by ananthakumaran** [comment]
> When writing files with JSX syntax, it would be ideal to increase the indent level at opening tags (`<div>`) and decrease the indent level at closing tags (`</div>`). Currently, all tags will end up at the same level, which hurt readability.

Did you try tide-format?

> How difficult would this be to implement? If it's a relatively big job, I'm happy to attempt to implement this myself, but would appreciate any pointers to the places in the current code that you think it might touch. I suspect that it might be best to implement as a new major mode that delegates most things to typescript-mode, as then it can be associated with the .tsx file extension, but again would appreciate any guidance you can provide!

There are two ways to fix it.

1) update/fix it in tsserver. TypeScript comes with tsserver(an editor service) which provides the formatting service. It provides a decent formatting for jsx, there are some edgecases and it's not much configurable.

2) handle it in typescript-mode. Currently typescript-mode is not aware of jsx/tsx syntax. I am not sure how much complex the implementation would be.

I use tide+web-mode for jsx/tsx files

**Comment by tbekolay** [comment]
> Did you try tide-format?

I was not aware of `tide-format`! This (or `web-mode`, which I've used in the past) works well enough. I wasn't aware that tide's flycheck worked with `web-mode` -- this is definitely sufficient for me right now. I see there's already an issue for adding jsx/tsx support to typescript-mode (https://github.com/ananthakumaran/typescript.el/issues/4). I'll look into that more to see if I can wrap my head around how to do it.

Thanks for the quick response! Closing this now as the alternatives above are nice, and the feature request has already been reported in the right place.
